gammasim/simtools
# Hdf5 simtel histograms

Closes #663.

### Summary
This PR addresses the issue described in #663. It reshapes and expands the `plot-simtel-histograms`. The following changes were implemented:
- I moved some functions from `generate-corsika-histograms` that would be needed in this tool to `simtools.utils.general`;
- The tool now supports exporting the histograms to one hdf5 file in addition to the pdf file;
- The structure of the tool was adapted inspired in the `generate-corsika-histograms` tool. With that, we maintain the standards (both in name and in structure, arguments, etc). This is important for the later stage, in which we will use a third tool to compare distributions (of both corsika and simtel_array histograms);
- The name of the tool also changed now, as it does not only plot histograms. The new name is in the same format as in `generate-corsika-histograms`: `generate-simtel-array-histograms`;
- Expanded the integration tests to test the additional functionalities

### To dos (completed)
- [x] Investigate whether we need to support all the available histograms or whether we can focus on the simulated events and use ctapipe containers as Karl suggested (https://github.com/gammasim/simtools/issues/663#issuecomment-1794981656)
Answer: We discussed in private and in the mattermost channel. We decided to support all the histograms as it is straightforward to pass them from the simtel output file to hdf5/pdf files. However, in a later stage, when developing the tool to compare the distributions, we will reconsider the possibility of using the structure available in ctapipe, as well as the weights we would like to give to each histogram.
- [x] Implement the new functionality of exporting to hdf5 in the application
- [x] Complement with the integration tests

### How to review
- As the name of the tool is different you might need to install simtools again after checking out the branch:
```pip install -e .```
- Run the following command and check the output (`test_hist_hdata.hdf5` and `test_hist_hdata.pdf`): ```simtools-plot-simtel-histograms --hist_file_names tests/resources/run2_gamma_za20deg_azm0deg-North-Prod5_test-production-5.hdata.zst --output_file_name test_hist_hdata --hdf5 --pdf```. To check `test_hist_hdata.hdf5` you might use the `read_hdf5` function as in `tests/unit_tests/simtel/test_simtel_histograms.py::test_export_histograms`
- Any other extra functionality that is not implemented here should be left to the following PRs. Please comment or open issues to register the ideas.
- Check the code and the module.
